molgenis-data-mapper/src/main/java/org/molgenis/data/mapper/service/impl/MappingServiceImpl.java

```diff
@@ -157,91 +154,129 @@ public MappingProject getMappingProject(String identifier)
+			}
 		}
 
 		try
 		{
 			LOG.info("Applying mappings to repository [" + targetMetaData.getName() + "]");
-			applyMappingsToRepositories(mappingTarget, targetRepo);
+			applyMappingsToRepositories(mappingTarget, targetRepo, addSourceAttribute);
+			if (hasSelfReferences(targetRepo.getEntityMetaData()))
+			{
+				LOG.info("Self reference found, applying the mapping for a second time to set references");
+				applyMappingsToRepositories(mappingTarget, targetRepo, addSourceAttribute);
+			}
 			LOG.info("Done applying mappings to repository [" + targetMetaData.getName() + "]");
 			return targetMetaData.getName();
 		}
 		catch (RuntimeException ex)
 		{
-			LOG.error("Error applying mappings, dropping created repository.", ex);
-			dataService.getMeta().deleteEntityMeta(targetMetaData.getName());
-			throw ex;
+			if (targetRepo.getName().equals(mappingTarget.getName()))
+			{
+				// Mapping to the target model, if something goes wrong we do not want to delete it
+				LOG.error("Error applying mappings to the target", ex);
+				throw ex;
+			}
+			else
+			{
+				// A new repository was created for mapping, so we can drop it if something went wrong
+				LOG.error("Error applying mappings, dropping created repository.", ex);
+				dataService.getMeta().deleteEntityMeta(targetMetaData.getName());
+				throw ex;
+			}
 		}
 	}
 
-	private void applyMappingsToRepositories(MappingTarget mappingTarget, Repository<Entity> targetRepo)
+	private void applyMappingsToRepositories(MappingTarget mappingTarget, Repository<Entity> targetRepo,
+			boolean addSourceAttribute)
 	{
 		for (EntityMapping sourceMapping : mappingTarget.getEntityMappings())
 		{
-			applyMappingToRepo(sourceMapping, targetRepo);
+			applyMappingToRepo(sourceMapping, targetRepo, addSourceAttribute);
 		}
 	}
 
-	private void applyMappingToRepo(EntityMapping sourceMapping, Repository<Entity> targetRepo)
+	private void applyMappingToRepo(EntityMapping sourceMapping, Repository<Entity> targetRepo,
+			boolean addSourceAttribute)
 	{
 		EntityMetaData targetMetaData = targetRepo.getEntityMetaData();
 		Repository<Entity> sourceRepo = dataService.getRepository(sourceMapping.getName());
 
-		// delete all target entities from this source
-		List<Entity> deleteEntities = targetRepo.query().eq("source", sourceRepo.getName()).findAll()
-				.filter(Objects::nonNull).collect(Collectors.toList());
-		targetRepo.delete(deleteEntities.stream());
-
-		Iterator<Entity> sourceEntities = sourceRepo.iterator();
-		List<Entity> mappedEntities = Lists.newArrayList();
-		while (sourceEntities.hasNext())
-		{
-			Entity sourceEntity = sourceEntities.next();
-			Entity mappedEntity = applyMappingToEntity(sourceMapping, sourceEntity, targetMetaData,
-					sourceMapping.getSourceEntityMetaData(), targetRepo);
-			mappedEntities.add(mappedEntity);
-
-			if (mappedEntities.size() == BATCH_SIZE || !sourceEntities.hasNext())
+		sourceRepo.iterator().forEachRemaining(sourceEntity -> {
```

Comment:
A TODO item maybe but sourceRepo.iterator() is hopelessly inefficient for large source repositories. forEachBatched is the method you need.